File: haskell/PerfectNumbers.hs
module PerfectNumbers (classify, Classification(..)) where

data Classification = Deficient | Perfect | Abundant deriving (Eq, Show)

listOfFactors :: Int -> [Int]
listOfFactors x = [ y | y <- [1..div x 2], mod x y == 0 ]

classify :: Int -> Maybe Classification
classify x
  | x <= 0 = Nothing
  | sumOfFactors < x = Just Deficient
  | sumOfFactors == x = Just Perfect
  | otherwise = Just Abundant
    where
      sumOfFactors = sum . listOfFactors $ x

File: javascript/perfect-numbers.js
export const classify = num => {
	if (num <= 0)
		throw new Error('Classification is only possible for natural numbers.');

	let sum = 0;

	for (let i = 1; i < num; i++) {
    if (num % i === 0)
      sum = sum + i;
	}

	return sum === num ? 'perfect' : sum > num ? 'abundant' : 'deficient';
};
